rust/examples/battery_service.rs

```diff
@@ -0,0 +1,187 @@
+        CharacteristicProperty,
+    },
+    hci::{packets::AddressType, Address},
+    transport::Transport,
+};
+use clap::Parser as _;
+use log::info;
+use owo_colors::OwoColorize;
+use pyo3::prelude::*;
+use rand::Rng;
+use std::time::Duration;
+
+#[pyo3_asyncio::tokio::main]
+async fn main() -> PyResult<()> {
+    env_logger::builder()
+        .filter_level(log::LevelFilter::Info)
+        .init();
+
+    let cli = Cli::parse();
+
+    let transport = Transport::open(cli.transport).await?;
+    let address = Address::from_be_hex("F0:F1:F2:F3:F4:F5", AddressType::RandomDeviceAddress)
+        .map_err(|e| anyhow!(e))?;
+    let device = Device::with_hci("Bumble", address, transport.source()?, transport.sink()?)?;
+    device.power_on().await?;
+
+    match cli.subcommand {
+        Subcommand::Client { target_addr } => {
+            let target = Address::from_be_hex(&target_addr, AddressType::RandomDeviceAddress)
+                .map_err(|e| anyhow!("{:?}", e))?;
+            run_client(device, target).await
+        }
+        Subcommand::Server => run_server(device).await,
+    }?;
+
+    // wait until user kills the process
+    tokio::signal::ctrl_c().await?;
+    Ok(())
+}
+
+async fn run_client(device: Device, target_addr: Address) -> anyhow::Result<()> {
+    info!("Connecting");
+    let conn = device.connect(&target_addr).await?;
+    let mut peer = Peer::new(conn)?;
+    info!("Discovering");
+    peer.discover_services().await?;
+    peer.discover_characteristics().await?;
+
+    let battery_service = peer
+        .create_service_proxy::<BatteryServiceProxy>()?
+        .ok_or(anyhow::anyhow!("No battery service found"))?;
+
+    info!("Getting characteristic");
+    let mut battery_level_char = battery_service
+        .battery_level()?
+        .ok_or(anyhow::anyhow!("No battery level characteristic"))?;
+    info!("Reading");
+    info!(
+        "{} {}",
+        "Initial Battery Level:".green(),
+        battery_level_char.read_value().await?
+    );
+    info!("Subscribing");
+    battery_level_char
+        .subscribe(|value| {
+            info!("{} {:?}", "Battery level update:".green(), value);
+        })
+        .await?;
+
+    Ok(())
+}
+
+async fn run_server(mut device: Device) -> anyhow::Result<()> {
+    let uuid = services::BATTERY.uuid();
+    let battery_level_uuid = Uuid16::from(0x2A19).into();
+    let battery_level = Characteristic::new(
+        battery_level_uuid,
+        CharacteristicProperty::Read | CharacteristicProperty::Notify,
+        AttributePermission::Readable.into(),
+        CharacteristicValueHandler::new(Box::new(BatteryRead), Box::new(NoOpWrite)),
+    );
+    let service = Service::new(uuid.into(), vec![battery_level]);
+    let service_handle = device.add_service(&service)?;
+
+    let mut builder = AdvertisementDataBuilder::new();
+    builder.append(CommonDataType::CompleteLocalName, "Bumble Battery")?;
+    builder.append(
```

Comment:
This is fairly manual, but given the use cases for Bumble that might want to do somewhat exciting things vis-a-vis spec compliance, I'm not sure it makes sense to make a more OS-style API that doesn't require you to assemble the advertisement and prevents you from doing things wrong...?

Reply:
Having a set of reusable AdvertismementData pre-set builders would be nice, for users who what just basic common stuff. That's something I've considered adding to the Python lib but haven't found the time to yet.